### PR TITLE
GDExtension: Add an interface for loading extra documentation

### PR DESCRIPTION
**Pull request by deralmas**
This is the culmination of quite a lot of attempts at interfacing with the editor documentation from GDExtension's API. It's not perfect but I think that it's finally usable enough considering the way that the doc system currently works.

GDExtensions can call `editor_help_load_xml_from_utf8_chars` or `editor_help_load_xml_from_utf8_chars_and_len` to register new "external" doc data in the form of an XML-formatted UTF-8 buffer. This should work with both workflows of using multiple calls of smaller data or a single call with everything inside. A single class' data will be automatically cleared when calling `classdb_unregister_extension_class`.

This is done through an extra documentation container in `EditorHelp` that stores a parsed copy of everything added/removed which is then appended to the main object when needed so that it can persist after a full doc regeneration. I can't think of a cleaner way with the current state of the doc system.

Some new methods got added to assist in this, namely: `DocTools::load_xml`, `EditorHelp::load_xml_buffer` and `EditorHelp::remove_doc`.

I'm not very experienced with GDExt's API conventions so feedback on it is appreciated.

A test extension is available over https://github.com/Riteo/threen/tree/doc (includes already an api dump, just call `scons gdextension_dir=gdextension`)

Based on the work of @akien-mga, supersedes #75415.

_Production edit: closes godotengine/godot-roadmap#61_

### CODE REVIEW

File: core/extension/gdextension_interface.h
Comment (dsnopek) [inline core/extension/gdextension_interface.h:2622]: This'll need to be `4.3` at the earliest:

```suggestion
 * @since 4.3
```

File: core/extension/gdextension_interface.h
Comment (dsnopek) [inline core/extension/gdextension_interface.h:2634]: Here too

File: core/extension/gdextension_manager.cpp
Comment (dsnopek) [inline core/extension/gdextension_manager.cpp:283]: Is this used anywhere? I don't see it

File: core/extension/gdextension_interface.cpp
Comment (dsnopek) [inline core/extension/gdextension_interface.cpp:1382]: There is going to be one tricky thing to solve here: code in `core/` is not supposed to directly call code in `editor/`.

We've solved this in a couple different ways in the past. For example, we couldn't directly call `EditorNode::add_plugin()` to register editor plugins. Instead, `EditorNode` sets some function pointers on `GDExtensionEditorPlugin`, so then `gdextension_interface.cpp` can call them.

Another example was with extension reloading: after it's finished, we need to refresh the editor inspector. So, we added the `extensions_reloaded` signal on `GDExtensionManager`, and `EditorNode` connects to that signal.

I'm not totally sure what the right thing to do in this case is, though. I don't think we'd want to do a signal, because we don't want scripts to be able to connect to it. So, maybe a solution closer to my first example would work best? Maybe we could have a static class named like `GDExtensionEditorHelp` which has some function pointers that the real `EditorHelp` could set when it's initialized?

**Review by dsnopek — COMMENTED** [review]
Thanks, this looks great! At a high-level, I think this is working exactly the right way :-)

There is only one more tricky thing we need to solve here, which I describe in one of my code comments.

File: core/extension/gdextension_interface.cpp
Comment (dsnopek) [reply core/extension/gdextension_interface.cpp:1382]: > Out of curiosity, why's this exactly? To avoid some sort of logical circular dependency, for cleanliness or something else?

I think for avoiding circular dependencies? We may even have this rule written down in the documentation somewhere but I couldn't find it just now. But the principle is that "higher level" things can depend on "lower level" things, but not the other way around. It goes something like Editor -> Scene -> Servers -> Core, so parts can depend on things to their right but not their left.

> Edit: oh I'm noticing that it's not building properly in the CI. I suppose it's related to that?

Possibly? But you may just be missing an `#include`

**Comment by dsnopek** [comment]
Thanks! The latest updates look great to me - including the stuff you were worried about in the 2nd commit! Just looking at the code, I don't have any more notes. :-) Personally, I think it'd be fine to squash the commits and take out of "draft" so that it can get some wider review.

The only thing left from my end is that I'd like to do some testing before hitting the "Approve" button.

(There will also be a non-trivial amount of work figuring out how best to integrate this functionality into godot-cpp, but that doesn't need to hold up this PR.)

**Comment by DmitriySalnikov** [comment]
Can it be merged before the 4.2 release?
It would be really cool to finally add documentation inside libraries.

**Comment by akien-mga** [comment]
No, 4.2 is in feature freeze and even at RC stage, it's not the time to add new features.

**Comment by dsnopek** [comment]
Sorry for taking so long to get back to this - I'm going to try to test this soon, hopefully this week. But just wanted to point out that this needs a rebase now.

**Review by dsnopek — COMMENTED** [review]
Alright!

I did my own local rebase and tested this. With the modification I describe in one of my review comments (about moving some code into `EditorHelp::_gen_extension_docs()`), it works great! :tada: 

In case it's helpful, I pushed my rebase to [this branch in my fork](https://github.com/dsnopek/godot/tree/pr/83747-rebased). I made the one functional change from my review there, but I didn't make the renaming changes.

And draft PR https://github.com/godotengine/godot-cpp/pull/1374 for godot-cpp can be used for testing. We'll need to do a bunch of design work and iteration to evolve it into the right solution for godot-cpp, but that shouldn't hold up this PR.

In any case, after the rebase and the renames, I'd be happy to approve this :-)

**Review by dsnopek — APPROVED** [review]
The code changes look great! I updated the changes in the draft PR for godot-cpp, and re-tested, and it worked great too.

IMO, this one is ready to merge :-)

**Comment by mihe** [comment]
I'm a bit late to the party here, but I've toyed around with this functionality for a bit now, and I'm starting to realize that you can't actually document project settings added by extensions when using this.

I'm assuming this is due to the fact that project settings are effectively documented as properties on the `ProjectSettings` class, which presumably can't be declared (and ideally merged) multiple times? I've tried naively declaring it in my extension anyway, but that seems to just get ignored.

Even if that did work I guess it wouldn't play very nice with `--doctool ../ --gdextension-docs` since that file would get nuked every time.

Any thoughts/advice?

**Comment by dsnopek** [comment]
Hm. Yeah, we may need a special way to handle project settings.

Having a `ProjectSettings.xml` that is merged into the existing properties isn't a terrible way to handle it. If that could be made to work, then maybe we just need an additional argument that can be used with `--doctool --gdextension-docs` to list which project settings you want exported?

**Comment by mihe** [comment]
> then maybe we just need an additional argument that can be used with `--doctool --gdextension-docs` to list which project settings you want exported?

Could that maybe be inferred from the property info of those project settings instead, seeing as how that's a `Dictionary` already, by adding some new boolean entry to it, like `"origin": "extension"` or something? It would be nice not to have to pass them to `--doctool` manually.

**Comment by dsnopek** [comment]
> Could that maybe be inferred from the property info of those project settings instead, seeing as how that's a `Dictionary` already, by adding some new entry to it, like `"origin": "extension"` or something?

Perhaps? It gets converted to a `PropertyInfo` struct internally, but maybe this info could be kept on the `ProjectSettings` singleton and the doc tool could get it from there?

It also occurs to me that we have the same problem with export settings, which are treated as properties of the export platform (ex `EditorExportPlatformAndroid`).

And, we have the same problem with GDScript based editor plugins too: they can also add project settings and export settings, and it would be nice to be able to document them.

So, maybe we need an API that can be used by either GDExtension or GDScript for this case? The documentation XML loading is GDExtension only.

**Comment by elementbound** [comment]
@mihe @dsnopek are you aware of any task or proposal about documenting custom ProjectSettings entries? I've recently ran into needing the same in https://github.com/foxssake/netfox/issues/244 and in https://github.com/foxssake/netfox/pull/314, both to document existing settings and to notify users that some settings are now deprecated and will be removed in the future.

**Pull request merged**
